ludwig/modules/serialization.py

```diff
@@ -0,0 +1,149 @@
+        child_group = group.create_group(k)
+        _save_state_to_group(child, child_group)
+
+
+def save_state_to_file(state: LudwigModuleState, url_or_path: str):
+    """Serializes LudwigModuleState to file."""
+    with upload_h5(url_or_path) as f:
+        # with h5py.File(f, "w") as f:
+        _save_state_to_group(state, f)
+
+
+def _load_state_from_group(group) -> LudwigModuleState:
+    """Restore LudwigModuleState from HDF5 group."""
+    return LudwigModuleState(
+        type=group.attrs["type"],
+        ludwig_version=group.attrs["ludwig_version"],
+        config=json.loads(group.attrs["config"]),
+        metadata=json.loads(group.attrs["metadata"]),
+        saved_weights={k: v[()] for k, v in group.items() if isinstance(v, h5py.Dataset)},
+        children={k: _load_state_from_group(v) for k, v in group.items() if isinstance(v, h5py.Group)},
+    )
+
+
+@lru_cache(maxsize=3)
+def load_state_from_file(path_or_uri: str) -> LudwigModuleState:
+    """Loads Ludwig Module state from a file."""
+    with download_h5(path_or_uri) as f:
+        # with h5py.File(f, "r") as f:
+        # The file object does double duty as the HDF5 root group
+        state = _load_state_from_group(f)
+    return update_state_for_current_version(state)
+
+
+def instantiate_module_from_state(state: LudwigModuleState, device: str = None) -> LudwigModule:
+    """Instatiates a module by restoring from saved state."""
+    cls = module_registry.get(state.type)
+    if cls is None:
+        logger.error(f"No Ludwig Module registered for name {state.type}")
+        return None
+    if not (hasattr(cls, "restore_from_state") and callable(cls.restore_from_state)):
+        logger.error(f"The @classmethod restore_from_state must be implemented to restore {state.type} objects")
+        return None
+    restored_module = cls.restore_from_state(state)
+    if restored_module is not None and device is not None:
+        restored_module = restored_module.to(device)
+    return restored_module
+
+
+def save(object: LudwigModule, path_or_uri: str, metadata: Dict[str, Any] = None):
+    """Saves Ludwig object to file or buffer."""
+    object_state = object.get_state(metadata=metadata)
+    save_state_to_file(object_state, path_or_uri)
+
+
+def load(path_or_uri: str, device: str = None) -> LudwigModule:
+    """Loads saved Ludwig module from file or buffer. If the module has parameters which are torch Tensors, device
+    specifies the device where tensors will be instantiated.
+
+    Args:
+        f: The file path or object to load from.
+        device: 'cuda' or 'cpu'
+    """
+    state = load_state_from_file(path_or_uri)
+    return instantiate_module_from_state(state, device=device)
+
+
+def update_state_for_current_version(state: LudwigModuleState) -> LudwigModuleState:
+    """Update saved Ludwig object from previous version."""
+    if state == LUDWIG_VERSION:
+        return state
+    else:
+        # TODO: check version, apply updates if needed.
```

Comment:
We certainly don't need to solve this now, though one proposal for how this could be implemented like for if/when parameters are renamed:

In ParameterMetadata, we can add a field i.e, `backwards_compatibility_source`, that contains tuple of older Ludwig version and which field that a module parameter should try to source its value from, if it's not explicitly in the config, as well as whether loading should crash if loading is not successful.

Hypothetical example: `output_size`'s metadata: `backwards_compatibility_source: [("<0.4", "fc_size", True)]`

Reply:
To add on to that, I'd like to add an option for a general hook to register a transformation (for a specific version and section of the config).  My plan is to implement that for regular configs (refactoring `backwards_compatibility.py`), then re-using that same system here in `serialization.py`.